src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumChest.java

```diff
@@ -214,22 +218,35 @@ public void addToolUsages(ItemStack stack, @Nullable World world, List<String> t
     protected void initializeInventory() {
         super.initializeInventory();
         this.itemInventory = new QuantumChestItemHandler();
-        this.outputItemInventory = new ItemHandlerProxy(new GTItemStackHandler(this, 0), exportItems);
         List<IItemHandler> temp = new ArrayList<>();
-        temp.add(outputItemInventory);
-        temp.add(itemInventory);
-        combinedInventory = new ItemHandlerList(temp);
+        temp.add(this.exportItems);
+        temp.add(this.itemInventory);
+        this.combinedInventory = new ItemHandlerList(temp);
+        this.outputItemInventory = new ItemHandlerProxy(new GTItemStackHandler(this, 0), combinedInventory);
 
     }
 
     @Override
     protected IItemHandlerModifiable createImportItemHandler() {
         return new GTItemStackHandler(this, 1) {
+
+            @Nonnull
+            @Override
+            public ItemStack insertItem(int slot, @Nonnull ItemStack stack, boolean simulate) {
+                if (!isItemValid(slot, stack)) return stack;
+                return GTTransferUtils.insertItem(getCombinedInventory(), stack, simulate);
+            }
+
             @Override
             public boolean isItemValid(int slot, @Nonnull ItemStack stack) {
                 NBTTagCompound compound = stack.getTagCompound();
+                ItemStack outStack = getExportItems().getStackInSlot(0);
+                boolean outStackMatch = true;
+                if (!outStack.isEmpty()) {
+                    outStackMatch = areItemStackIdentical(stack, outStack);
+                }
                 if (compound == null) return true;
-                return !(compound.hasKey(NBT_ITEMSTACK, NBT.TAG_COMPOUND) || compound.hasKey("Fluid", NBT.TAG_COMPOUND)); //prevents inserting items with NBT to the Quantum Chest
+                return !(compound.hasKey(NBT_ITEMSTACK, NBT.TAG_COMPOUND) || compound.hasKey("Fluid", NBT.TAG_COMPOUND)) && outStackMatch; //prevents inserting items with NBT to the Quantum Chest
```

Comment:
The `outStackMatch` check should be moved before the nbt checks, to allow for short-circuiting.